### ISSUE TITLE
Creating Pub/Sub PublisherClient prevents application from terminating (?!?)

### ISSUE THREAD

**Issue by chrsmith** [GoogleCloudPlatform/gcloud-dotnet#186]
I'll admit the following issue sounds odd. So odd that I wouldn't be surprised if it was just an issue with my computer or its configuration. But please try the simple repo for yourself and see if I am crazy or not.

Running the following application won't terminate. It will print "Done" to the console, but the process will remain active even after `Main` has exited. Commenting out the instantiation of the `PublisherClient` will cause the program to immediately terminate as you would expect.

Perhaps a hang in the finalizer?

Repro:
````
using Google.Pubsub.V1;
using System;

namespace BugRepro
{
    /// <summary>
    /// Bug repro. Requires the following NuGet dependencies (from MyGet)
    /// 
    /// Install-Package Google.PubSub.V1 -Pre
    /// </summary>
    class Program
    {
        static void Main(string[] args)
        {
            // Application will hang after Main() exits. (Presumably some GRPC goo in the
            // background is preventing the application from terminating?)
            // Commenting out this line of code will cause the application to immediately
            // shutdown. (As expected.)
            PublisherClient publisherClient = PublisherClient.Create();
            Console.WriteLine("Done");
        }
    }
}
````

**Comment by jskeet** [comment]
Will try to reproduce. I wonder whether there's some bug around channels which have been created but not connected to anything yet.

**Comment by jskeet** [comment]
Okay, reproduced and confirmed that calling `ShutdownDefaultChannelsAsync` fixes the problem.

I'm still not sure why we need foreground threads in gRPC, which is what's causing the problem. Will talk to gRPC friends...

**Comment by jtattermusch** [comment]
Some rationale why foreground threads in gRPC thread pool:
1. Upon shutdown you probably want to clean up the state. This can be done either by user triggering the shutdown manually, by some kind of shutdown hook (no support for that from gRPC itself), or by relying on the finalizer do dispose disposable objects.
2. gRPC stack is multithreaded by nature and having finalizer thread to clean up the state is problematic (because one shouldn't do any sychronization on the finalizer thread).  The foreground threads are there to make sure that finalizer doesn't attempt to clean up something that can be impossible to cleanup yet (might result in a crash). Once the multi-threaded portion of gRPC has been shutdown (by shutting down all channels and servers), finalizer will work correctly.

**Comment by jtattermusch** [comment]
https://github.com/grpc/grpc/pull/6754

**Comment by jskeet** [comment]
Fixed by grpc 0.15.
